### ISSUE TITLE
[discussion] Message queuing is dangerous and unreliable

### ISSUE THREAD

**Issue by zikaari** [zikaari/webext-bridge#23]
While I was once proud of this feature when I implemented it, I have developed a counter opinion that message queuing is not safe _(especially now that the background service workers, that webext-bridge relies on, terminate)_.

Take this example:

A content script calls `sendMessage` to have a message delivered to `devtools`. Because the devtools were not open, the message gets queued in the service worker context for a later delivery. Keep in mind that `sendMessage` call returned a Promise that's still pending:
   - Outcome A: The service worker terminates, the queue is lost, and content script keeps waiting on Promise that is never going to fulfill.
   - Outcome B: The user opens devtools, but AFTER the user navigates to another webpage in the same tab, now the devtools is receiving the message that was sent in the "past" and is probably no longer relevant.

Perhaps we need to get rid of message queuing altogether, or perhaps the problem is that we do indefinite message queuing with no timeouts and such restraints.

I want to hear your opinions on this matter so we can reach a consensus.

Learn more about service worker termination [here](https://bugs.chromium.org/p/chromium/issues/detail?id=1152255)

**Comment by smeijer** [comment]
Thanks for the ping! I appreciate the notice.

What downsides do you see for simply adding a timeout?

And how about option 3 -  Flush the queue and cancel the promises on navigation? Would that be an option?

**Comment by zikaari** [comment]
Even with timeouts in place, Outcome B is still a possibility.

I guess a proper approach might be to have all runtime contexts be aware of the state of the "mesh". `content-script` needs to be made aware that the service worker has been terminated and it's end of story for all pending `sendMessage`s. `devtools` shouldn't receive a message if it's from the "past", from the previous content-script that was running in the same tab. It might be quite complex to implement but I think it's necessary for the sake of an extension's data and state integrity.

**Comment by p6l-richard** [comment]
Hi @zikaari,

I found this issue today after I wanted to opt-out of this queueing behaviour.

I'm outlining my case here as it seems fitting to collect use cases. 
Note: My use case is currently development only, I haven't tested this specific behaviour on a chrome released extension yet. 

Context:
I'm using [Plasmo](https://www.plasmo.com/) do build my web app and it uses parcel's extension HMR under the hood. 

Problem: 
I observed that whenever the HMR is triggered by me saving a file, the behaviour of sendMessage is off: I don't see the messages delivered anymore.
My initial thought is that I ended up with your above described first scenario:
> - Outcome A: The service worker terminates, the queue is lost, and content script keeps waiting on Promise that is never going to fulfill.

Then, I refreshed the page in my browser and all of a sudden I get multiple messages delivered from the clicks that I thought were lost.  
Concluding, I'm actually in the Outcome B described.
> - Outcome B: The user opens devtools, but AFTER the user navigates to another webpage in the same tab, now the devtools is receiving the message that was sent in the "past" and is probably no longer relevant.

Hence, here's the flow that I _think_ happens (happy for any corrections if you spot anything immediately):
- After reloading the extension, the content-script on an already open tab goes into `loading` mode
- I trigger the `onAction` in the background service worker
- Background service worker uses `sendMessage` API to deliver to content-script
- As the content-script is still in loading state, it queues the messages[1]
- I have to manually put the content-script out of the loading state by refreshing (or alternatively visiting a new page or similar) -> queue is processed with old events

Questions:
[1] Is there a way for me to hook into what's currently handled for the message queue and customise my behavior? E.g. in my case tracking and then dropping the message from the queue and rejecting the promise. My personal DX preference is to opt-out of queueing for these cases.
[2] I notice that with `webext-bridge` specifically I cannot check for the `chrome.runtime.lastError` -- is that because of the `webextension-polyfill` which deactivates that behaviour? What's a good way for me to understand if the message delivery failed?

**Comment by zikaari** [comment]
After some time, I have a feeling that **Outcome A** is just a theory and can never actually occur since service worker's termination is indefinitely prevented by `onDisconnect` listeners that create a new connection with the service worker, thereby preventing its suspension.

**Outcome B** is still valid and I've personally noticed in some of my own extensions. I'll fix this by implementing smarter message invalidation by removing them from the delivery queue once the sender itself is gone _(eg: content script unmounted)_.

Keep your eyes on the [`next`](https://github.com/zikaari/webext-bridge/tree/next) branch of this repo, the fix will land there.

**Comment by zikaari** [comment]
>After some time, I have a feeling that Outcome A is just a theory and can never actually occur since service worker's termination is indefinitely prevented by `onDisconnect` listeners that create a new connection with the service worker, thereby preventing its suspension.

This assessment was not true. The service worker DOES get terminated, `onDisconnect` traps just re-spawn a new one. Since a fresh instance is spawned all the state is reset and nothing gets carried over.

___

However, both the scenarios **Outcome A** and **Outcome B**, plus another case where the destination might disconnect before it could respond have now been fixed in https://github.com/zikaari/webext-bridge/pull/38. See the PR description to learn more about how its been done.
